src/highlevel.jl

```diff
@@ -660,18 +660,35 @@ end
 
 Read data from the device into the given buffer.
 """
-function Base.read!(s::Stream{T}, buffer::NTuple{N, AbstractVector{T}}; timeout=nothing) where {N, T}
+function Base.read!(s::Stream{T}, buffers::NTuple{N, AbstractVector{T}}; timeout=nothing) where {N, T}
+    t_start = time()
     timeout === nothing && (timeout = 0.1u"s") # Default from SoapySDR upstream
+    timeout_s = uconvert(u"s", timeout).val
+    timeout_us = uconvert(u"μs", timeout).val
 
     total_nread = 0
-    to_read_ct = length(first(buffer)) # TODO assert all equal
-    while total_nread < to_read_ct
-        nread, flags, timens = SoapySDRDevice_readStream(s.d, s, Ref(map(b -> pointer(b, total_nread+1), buffer)), to_read_ct-total_nread, uconvert(u"μs", timeout).val)
-        timens = timens * u"ns"
+    if !all([length(b) for b in buffers] .== length(first(buffers)))
```

Comment:
Is the compiler smart enough to make is non allocating?

Reply:
No, but I can fix that.